core/services/offchainreporting/arbitrum_block_translator.go

```diff
@@ -0,0 +1,259 @@
+	mark := time.Now()
+	var n int
+	defer func() {
+		duration := time.Since(mark)
+		logger.Debugw(fmt.Sprintf("ArbitrumBlockTranslator#binarySearch completed in %s with %d total lookups", duration, n), "finishedIn", duration, "err", err, "nLookups", n)
+	}()
+	var h *models.Head
+
+	// l2lower...l2upper is the exclusive range of L2 block numbers in which
+	// transactions that called block.number will return the given L1 block
+	// number
+	var l2lower int64
+	var l2upper int64
+
+	var skipUpperBound bool
+
+	{
+		var maybeL2Upper *int64
+		l2lower, maybeL2Upper = a.reverseLookup(targetL1)
+		if maybeL2Upper != nil {
+			l2upper = *maybeL2Upper
+		} else {
+			// Initial query to get highest L1 and L2 numbers
+			h, err = a.ethClient.HeaderByNumber(ctx, nil)
+			n++
+			if err != nil {
+				return nil, nil, err
+			}
+			if h == nil {
+				return nil, nil, errors.New("got nil head")
+			}
+			if !h.L1BlockNumber.Valid {
+				return nil, nil, errors.New("head was missing L1 block number")
+			}
+			currentL1 := h.L1BlockNumber.Int64
+			currentL2 := h.Number
+
+			a.cachePut(currentL2, currentL1)
+
+			// NOTE: This case shouldn't ever happen but we ought to handle them in the least broken way possible
+			if targetL1 > currentL1 {
+				// real upper must always be nil, we can skip the upper limit part of the binary search
+				logger.Debugf("ArbitrumBlockTranslator#BinarySearch target of %d is above current L1 block number of %d, using nil for upper bound", targetL1, currentL1)
+				return big.NewInt(currentL2), nil, nil
+			} else if targetL1 == currentL1 {
+				// NOTE: If the latest seen L2 block corresponds to the target L1
+				// block, we have to leave the top end of the range open because future
+				// L2 blocks can be produced that would also match
+				skipUpperBound = true
+			}
+			l2upper = currentL2
+		}
+	}
+
+	logger.Tracef("ArbitrumBlockTranslator#BinarySearch starting search for L2 range wrapping L1 block number %d between bounds [%d, %d]", targetL1, l2lower, l2upper)
+
+	// LEFT EDGE
+
+	// Expect either left or right search to make an exact match, if they don't something has gone badly wrong
+	var exactMatch bool
+	{
+		l2lower, err = search(l2lower, l2upper+1, func(l2 int64) (bool, error) {
+			l1, miss, err2 := a.arbL2ToL1(ctx, l2)
+			if miss {
+				n++
+			}
+			if err2 != nil {
+				return false, err2
+			}
+			if targetL1 == l1 {
+				exactMatch = true
+			}
+			return l1 >= targetL1, nil
+		})
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	// RIGHT EDGE
+	if !skipUpperBound {
+		var r int64
+		r, err = search(l2lower, l2upper+1, func(l2 int64) (bool, error) {
+			l1, miss, err2 := a.arbL2ToL1(ctx, l2)
+			if miss {
+				n++
+			}
+			if err2 != nil {
+				return false, err2
+			}
+			if targetL1 == l1 {
+				exactMatch = true
+			}
+			return l1 > targetL1, nil
+		})
+		if err != nil {
+			return nil, nil, err
+		}
+		l2upper = r - 1
```

Comment:
It wasn't immediately clear to me how the bounds are found (I first thought the two closures are identical) -- I think this area could use a comment or two to explain: you're finding the left edge by looking for the value, then finding the right edge by looking for the value right after and subtracting one.